### ISSUE TITLE
Discuss decorating scope of `ServerBuilder.decorator(decorating)`

### ISSUE THREAD

**Issue by ikhoon** [line/armeria#2057]
While I was working on `Decorate a service by path mapping` #582, I realized that `ServerBuilder.decorator(decorator)` works for all virtual-host though `ServerBuilder.service(route, service)` only works for default virtual host.
```java
/**
 * Binds the specified {@link Service} at the specified {@link Route} of the default
 * {@link VirtualHost}.
 */
public ServerBuilder service(Route route, Service<HttpRequest, HttpResponse> service) {
    defaultVirtualHostBuilder.service(route, service);
    return this;
}
```

I'm going to add `ServerBuilder.decorator(route, service)` for decorating a service whose route matches the given route. `ServerBuilder` will have similar methods but they work on different scopes.

* `ServerBuilder.service(route, service)` binds service for **default** virtual host.
* `ServerBuilder.decorator(decorator)` decorates **all** virtual hosts.
* `ServerBuilder.decorator(route, decorator)` will decorate **all** virtual hosts. (new method)

Now we keep the original policy to avoid breaking change, but we need to discuss this policy whether decorating default virtual-host or all virtual-host for API consistency. 

The following code shows how `ServerBuilder.decorator(decorator)` decorates all virtual hosts.

````java
 ServerRule server = new ServerRule() {
    @Override
    protected void configure(ServerBuilder sb) throws Exception {
        sb.service("/", (ctx, req) -> HttpResponse.of(HttpStatus.OK))
          .service("/test", (ctx, req) -> HttpResponse.of(HttpStatus.OK))
          .decorator((delegate, ctx, req) -> {  // decorates all virtual-host
              ctx.addAdditionalResponseHeader("global_decorator", "true");
              return delegate.serve(ctx, req);
          })
          .virtualHost("*.example.com")
          .service("/", (ctx, req) -> HttpResponse.of(HttpStatus.OK))
          .decorator((delegate, ctx, req) -> {
              ctx.addAdditionalResponseHeader("virtualhost_decorator", "true");
              return delegate.serve(ctx, req);
          })
          .build();
```

**Comment by trustin** [comment]
I think I like the current behavior of `ServerBuilder.decorator()` and `ServerBuilder.server()` and `ServerBuilder.route()` should yield the same behavior. For example, if a user added a service like this:

```java
sb.virtualHost("*.foo.com").route().get("/foo")....build(...);
sb.route().get("/")...build(...);
```

.. the default virtual host would have service bindings for `/` only and the virtual host `*.foo.com` would have service bindings for both `/foo` and `/`.

**Comment by ikhoon** [comment]
> I think I like the current behavior of ServerBuilder.decorator() and ServerBuilder.server() and ServerBuilder.route() should yield the same behavior.

I agree with your opinion. I will try to bind the default virtual host service to a custom virtual host service in [Provide a way to decorate a service by `Route`](https://github.com/line/armeria/pull/2040). 

But if it needs to change a lot of code, I will make another PR for this issue.

**Comment by trustin** [comment]
Fixed via #2040
